ftp/src/main/java/ch/cyberduck/core/ftp/FTPSocket.java

```diff
@@ -0,0 +1,293 @@
+ * race conditions due to premature socket closure during FTP data connections.
+ * <p>
+ * This fixes issues where:
+ * - macOS: Intermittent "426 Connection closed; transfer aborted" errors due to socket closure before server ACKs
+ * - Windows: Intermittent transfer hanging due to socket closure before client sends FIN with last data packet
+ * <p>
+ * The proper TCP shutdown sequence is:
+ * 1. Call shutdownOutput() to send FIN while keeping socket input open for ACKs
+ * 2. Drain input to wait for ACKs until we receive server FIN
+ * 3. Close the socket to release resources
+ */
+public class FTPSocket extends Socket {
+    private static final Logger log = LogManager.getLogger(FTPSocket.class);
+
+    private final Socket delegate;
+
+    private final ConnectionTimeout connectionTimeoutPreferences = ConnectionTimeoutFactory.get();
+
+    public FTPSocket(final Socket delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            if(delegate.isClosed() || delegate.isOutputShutdown() || !delegate.isConnected()) {
+                log.debug("Socket already closed {}", delegate);
+            }
+            else {
+                // Shutdown output to send FIN, but keep socket open to receive ACKs
+                log.debug("Shutting down output for socket {}", delegate);
+                delegate.shutdownOutput();
+
+                // Wait for server FIN by draining any remaining data
+                // This ensures all our data packets are ACKed before closing
+                try {
+                    // Avoid hanging in case the server malfunctions
+                    delegate.setSoTimeout(connectionTimeoutPreferences.getTimeout() * 1000);
+                    InputStream in = delegate.getInputStream();
+                    byte[] buffer = new byte[8192];
+                    int bytesRead;
+                    // Read until EOF (server closes its side) or timeout
+                    while((bytesRead = in.read(buffer)) != -1) {
+                        log.debug("Drained {} bytes from socket after shutdown", bytesRead);
+                    }
+                    log.debug("Received EOF from server, all data acknowledged");
+                }
+                catch(java.net.SocketTimeoutException e) {
+                    // Timeout is acceptable - server may not close its side
+                    log.debug("Timeout waiting for server EOF, proceeding with close");
+                }
+                catch(IOException e) {
+                    // Other errors during drain are acceptable
+                    log.debug("Error draining socket: {}", e.getMessage());
+                }
+            }
+        }
+        catch(IOException e) {
+            log.warn("Failed to shutdown output for socket {}: {}", delegate, e.getMessage());
+        }
+        finally {
+            log.debug("Closing socket {}", delegate);
+            delegate.close();
```

Comment:
I noticed that at least on macOS, with 2/3 of the FTP servers I'm testing, there's an exact 1,000ms delay when `delegate.close()` is called after a file upload. I tried a bunch of things, including SO_LINGERING(true/false), NO_DELAY, SO_TIMEOUT(1), and nothing helps. Any ideas?

Reply:
Tested on Windows, there's no delay with either 3 of the FTP servers. So this is an isolated issue with macOS it seems.